pydantic/_internal/_generate_schema.py

```diff
@@ -210,6 +211,42 @@ def modify_model_json_schema(
     return json_schema
 
 
+JsonEncoders = Dict[Type[Any], JsonEncoder]
+
+
+def _add_custom_serialization_from_json_encoders(
+    json_encoders: JsonEncoders | None, tp: Any, schema: CoreSchema
+) -> CoreSchema:
+    """Iterate over the json_encoders and add the first matching encoder to the schema.
+
+    Args:
+        json_encoders (JsonEncoders | None): A dictionary of types and their encoder functions.
+        tp (Any): The type to check for a matching encoder.
+        schema (CoreSchema): The schema to add the encoder to.
+    """
+    if json_encoders is None:
+        return schema
+    if 'serialization' in schema:
+        return schema
+    # Check the class type and its superclasses for a matching encoder
+    # Decimal.__class__.__mro__ (and probably other cases) doesn't include Decimal itself
+    for base in (tp, *tp.__class__.__mro__[:-1]):
+        try:
+            encoder = json_encoders[base]
+        except KeyError:
+            continue
```

Comment:
```suggestion
        encoder = json_encoders.get(base)
        if encoder is None:
            continue
```
fewer loc and since I think it should be faster since you'll get a KeyError more often than not (my understanding is that using `try: except:` is faster when you don't hit the exception, and slower when you do). I have no problem keeping it as-is if you prefer though.

Reply:
Okay yeah, I was curious so actually tried, and the `.get` seems to be ~5x faster:
```python
import time

a = dict(enumerate(range(100)))
n = 10_000_000

t0 = time.time()
for _ in range(n):
    try:
        a[100]
    except KeyError:
        continue
t1 = time.time()
print(f'{t1 - t0:.3f}')
#> 2.540

t0 = time.time()
for _ in range(n):
    x = a.get(100)
    if x is None:
        continue
t1 = time.time()
print(f'{t1 - t0:.3f}')
#> 0.646
```

Reply:
And if you change it to `a[99]` and `a.get(99)` instead of 100, the `try: except:` is ~2x faster, so, intuition confirmed.

I also confirmed these performance ratios hold if you make the dict have just a single item.

Reply:
Ok just keep in mind that this was copy pasta from V1